### PR TITLE
refactor get clients

### PR DESCRIPTION
**Pull request by haiyang1985**
## What is the purpose of the change

XXXXX

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/DubboProtocol.java [hunk 1]
```diff
@@ -396,12 +396,56 @@ private void optimizeSerialization(URL url) throws RpcException {
         optimizeSerialization(url);
 
         // create rpc invoker.
-        DubboInvoker<T> invoker = new DubboInvoker<T>(serviceType, url, getClients(url), invokers);
+        DubboInvoker<T> invoker = new DubboInvoker<T>(serviceType, url, getClientsV2(url), invokers);
```
Comment (lzf971107) [inline dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/DubboProtocol.java]: After you refactored your code, your unit tests failed. Please fix these exceptions.

[ERROR] DubboLazyConnectTest.testSticky1:53 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] DubboLazyConnectTest.testSticky3:67 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] ExplicitCallbackTest.TestCallbackConsumerLimit:176 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] ExplicitCallbackTest.TestCallbackProviderLimit:203 Unexpected exception type thrown ==> expected: <org.apache.dubbo.rpc.RpcException> but was: <java.lang.IndexOutOfBoundsException>
[ERROR] Errors:
[ERROR] DubboInvokerAvilableTest.test_Lazy_ChannelReadOnly:131 » IndexOutOfBounds Inde...
[ERROR] DubboInvokerAvilableTest.test_NoInvokers:118 » IndexOutOfBounds Index: 0, Size...
[ERROR] DubboInvokerAvilableTest.test_Normal_ChannelReadOnly:78 » IndexOutOfBounds Ind...
[ERROR] DubboInvokerAvilableTest.test_Normal_available:67 » IndexOutOfBounds Index: 0,...
[ERROR] DubboLazyConnectTest.testSticky2:62 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboLazyConnectTest.testSticky4:81 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testDemoProtocol:62 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testDubboProtocol:71 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testDubboProtocolMultiService:148 » IndexOutOfBounds Index: ...
[ERROR] DubboProtocolTest.testDubboProtocolWithMina:104 » IndexOutOfBounds Index: 0, S...
[ERROR] DubboProtocolTest.testNonSerializedParameter:178 » IndexOutOfBounds Index: 0, ...
[ERROR] DubboProtocolTest.testPerm:166 » IndexOutOfBounds Index: 0, Size: 0
[ERROR] DubboProtocolTest.testRemoteApplicationName:208 » IndexOutOfBounds Index: 0, S...
[ERROR] DubboProtocolTest.testReturnNonSerialized:192 » IndexOutOfBounds Index: 0, Siz...
[ERROR] ExplicitCallbackTest.TestCallbackMultiInstans:128->initOrResetService:90->referService:65 » IndexOutOfBounds
[ERROR] ExplicitCallbackTest.TestCallbackNormal:107->initOrResetService:90->referService:65 » IndexOutOfBounds
[ERROR] MultiThreadTest.testDubboMultiThreadInvoke:51 » IndexOutOfBounds Index: 0, Siz...
[ERROR] ReferenceCountExchangeClientTest.test_counter_error:156->init:223 » Arithmetic
[ERROR] ReferenceCountExchangeClientTest.test_mult_share_connect:117->init:223 » Arithmetic
[ERROR] ReferenceCountExchangeClientTest.test_multi_destory:139->init:223 » Arithmetic
[ERROR] ReferenceCountExchangeClientTest.test_not_share_connect:103->init:221->referInvoker:72 » IndexOutOfBounds
[ERROR] ReferenceCountExchangeClientTest.test_share_connect:92->init:223 » Arithmetic ...
[ERROR] RpcFilterTest.testRpcFilter:48 » IndexOutOfBounds Index: 0, Size: 0

@haiyang1985

Comment (haiyang1985) [reply dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/DubboProtocol.java]: Current PR is for discuss only, not my final version.

**Comment by AlbumenJ** [comment]
@haiyang1985 Hi, thanks for your contributions.

Was this pr still useful for the latest master branch?

**Comment by haiyang1985** [comment]
> @haiyang1985 Hi, thanks for your contributions.
> 
> Was this pr still useful for the latest master branch?

No, it's not the final version. This pr is for discuss only, it should be closed.

**Pull request closed without merging**